### PR TITLE
build jellyfin from source, jellyfin: 10.7.2 -> 10.7.5

### PR DESCRIPTION
**Pull request by jojosch**
###### Motivation for this change

Added a "from source" build for `jellyfin` and `jellyfin-web` (this one can be reused for the `jellyfin-media-player` when the changes of the [current fork](https://github.com/NixOS/nixpkgs/blob/92c77733acdb7acb71c9201fe4d445afb2949b1a/pkgs/applications/video/jellyfin-media-player/default.nix#L38-L41) are getting upstreamed).

I played some files (with and without transcoding), added some of the available plugins and they seem to be working fine. ~~I also ran a new server with the "old" 10.5, updated that to the "from-source" 10.5 and then to the "from-source" 10.7 without any major problems (the server needed to be restarted once after the 10.5 -> 10.7 update after the db migrations had been run).~~

Due to not having any machines with `arm` i wasn't able to test the build on this architecture.

Also added an update script for both `jellyfin` and `jellyfin-web`.

I added the current jellyfin maintainers to the new `jellyfin-web` package - if that's not OK for one of you i'll remove you.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
```
1 package added:
jellyfin-web (init at 10.7.5)

1 package updated:
jellyfin (10.7.2 → 10.7.5)

2 packages built:
jellyfin jellyfin-web
```
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
```bash
# NixOS
/nix/store/9cy4sbbmdr9zclca0q23sc2dk1adsxkn-jellyfin-10.7.2	  558501584
/nix/store/0dalbc9dkv855p9gdxfixp4mpxi5j7i6-jellyfin-10.7.2	  492186376 (from source)
/nix/store/91asbi4rd4inrkh05a1yrdiji6j1ziqf-jellyfin-web-10.7.2	   48892064
/nix/store/98n8137bvb75z6dqhc6qc2ifpbrkhilr-jellyfin-10.7.5	  491992656
/nix/store/lpflmvn94rnl1n160354cnrqd2p8cprp-jellyfin-web-10.7.5	   48891944

# macOS
/nix/store/spcf0i4pr3dd1jza9g6rr93kz49kihgz-jellyfin-10.7.2	  533905472
/nix/store/qy59h5zzc07v3rmqajrxanbq35kijzgw-jellyfin-10.7.2	  373582688 (from source)
/nix/store/wqkf82pnm2qnxagnbkpq20grbzj5ncp9-jellyfin-web-10.7.2	   48892216
```
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Church-** [comment]
Have tested this out on a few boxes at home, checks out so far.

File: pkgs/servers/jellyfin/server-generic.nix [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+        name = "$pkg_name";
+        version = "$pkg_version";
+        sha256 = "$pkg_sha256";
+      })
+    EOF
+    done < <(find ./nuget_tmp.packages -name '*.nuspec' | sort)
+    echo "]" >>"$deps_file"
+
+    popd
+    rm -r "$src"
```
Comment (SuperSandro2000) [inline pkgs/servers/jellyfin/server-generic.nix:206]: Can we move this into a shell file to get proper highlighting, formatting and easier shellcheck support?

**Review by SuperSandro2000 — COMMENTED** [review]
Can we reuse the global nodePackages file somehow?

File: pkgs/servers/jellyfin/server-generic.nix (hunk 1 above)
Comment (jojosch) [reply pkgs/servers/jellyfin/server-generic.nix]: I've moved the update logic to separate shell scripts. Also fixed some warnings from `shellcheck`.

I've added some arguments for the needed variables, which i pass from the `passthru.updateScript` to the shell scripts.

**Comment by minijackson** [comment]
Thank you for this! I think having a built from source Jellyfin is much better than downloading official releases.

A couple of concerns, though:

1. I know next to nothing in C#, so I fear that the package is going to surpass my knowledge. I'll still help however I can, though.

2. I think adding the jellyfin 10.5 variant was a mistake on my part in retrospect, sorry about that.

The reason is that Jellyfin doesn't have a 10.5 release channel, so this is just a sitting unsupported package.

What's more is that there recently was a security advisory for Jellyfin <10.7.0, so I think we really should be removing that, and backport the latest Jellyfin. I'm not saying this is the PR that should be backported, since it's still a draft, but I'm just opening the discussion to see if anyone has concerns. The original reason why I suggested keeping 10.5 is that upstream mentioned in their Changelog that they made some incompatible changes in their database layout, and recommended backuping the database, and I wanted a way for end-users to have a way to do that before upgrading.

But for NixOS master, I think we should definitely remove the 10.5 variant, it should also simplify some things in this PR.

3. Not really an issue in this PR per-se, but I think a lot of the build process could be automated: I think we should have a C# builder. I see a lot of duplicated code over a couple of Nix packages (ryujinx, osu-lazer, WasabiBackend, ...), and I think this is something that could gain a lot by getting central attention. 

4. It seems the Mono segfaulted for the aarch64 architecture. Can anyone reproduce this? Should we put the package as broken for aarch64?

Otherwise, nix-review builds fine, test executes fine, and quick testing shows that the server seems to run great :tada:

**Comment by jojosch** [comment]
> But for NixOS master, I think we should definitely remove the 10.5 variant, it should also simplify some things in this PR.

That would indeed simplify the build scripts a fair bit :smile:

I would then just add the from-source build for the "latest" jellyfin and leave the 10.5 as-is (for now).

**Comment by purcell** [comment]
Heads up that #120520 will remove Jellyfin 10.5.0, so you might find this can be further simplified.

File: pkgs/servers/jellyfin/update.sh [hunk 2]
```diff
@@ -0,0 +1,52 @@
+#!/usr/bin/env nix-shell
+#!nix-shell -i bash -p curl jq common-updater-scripts dotnetCorePackages.sdk_5_0 gnused nix coreutils findutils
+
+set -u -e -o pipefail
```
Comment (SuperSandro2000) [inline pkgs/servers/jellyfin/update.sh:4]:
```suggestion
set -euo pipefail
```

**Comment by jojosch** [comment]
I've rebased on master to get the enhanced jellyfin-tests (#120548) - they run successfully.

**Comment by jojosch** [comment]
Also updated to the latest release https://github.com/jellyfin/jellyfin/releases/tag/v10.7.5.

10.7.3 resolves https://github.com/jellyfin/jellyfin/security/advisories/GHSA-rgjw-4fwc-9v96

**Comment by AkechiShiro** [comment]
To use this package when it is fully merged inside the Nixpkgs repo, will there be the need to alter the configuration of Jellyfin inside the nix files ? (I know that the answer to this question, depends on the configuration file however in my case, I'm under the assumption that I'm using the latest unstable Jellyfin from the unstable Nixpkgs repo).

Like this : 
```nix
{config, pkgs,  ...}:
let pkgUnstable = import <nixos-unstable> { config = { allowUnfree = true; }; };
in {
    services.jellyfin.package = pkgsUnstable.jellyfin;
    ...
}
```

**Comment by jojosch** [comment]
That'll pick up the updated package as expected - no further changes needed.

I'm using flakes and added my branch as another input until it's merged and I did set the equivalent `services.jellyfin.package = MY_INPUT.jellyfin;`

**Comment by minijackson** [comment]
@ofborg test jellyfin

**Comment by Kranzes** [comment]
How is this PR going so far?

**Review by minijackson — APPROVED** [review]

**Comment by jonringer** [comment]
Arch failure looks related to https://github.com/NixOS/nix/issues/4810

**Comment by jonringer** [comment]
We need a better solution for node files than massive files. Similar to nuget

**Review by jonringer — APPROVED** [review]
LGTM

Result of nixpkgs-review pr 120344 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)<details>  <summary>1 package blacklisted:</summary>  <ul>    <li>tests.nixos-functions.nixos-test</li>  </ul></details><details>  <summary>2 packages built:</summary>  <ul>    <li>jellyfin</li>    <li>jellyfin-web</li>  </ul></details>

**Comment by kevincox** [comment]
This broke my setup because I was assuming the location of jellyfin-web was inside the jellyfin package. However I am fine with that. I just needed to make a small change to my config and the result is cleaner.

```patch
 			location /web/ {
-				alias ${pkgs.jellyfin}/opt/jellyfin/jellyfin-web/;
+				alias ${pkgs.jellyfin-web}/share/jellyfin-web/;
```

**Pull request merged**